pkg/crds/crds.go

```diff
@@ -26,142 +28,120 @@ import (
+func (r *CRDReconciler) LoadCRDs() error {
+	for _, file := range []string{"crd-10.yaml", "crd-11.yaml", "crd-14.yaml"} {
+		f, err := istio_crds.CRDs.Open(file)
+		if err != nil {
+			return err
+		}
+
+		err = r.load(f)
+		if err != nil {
+			return err
+		}
 	}
-}
 
-func MeshPolicy() *extensionsobj.CustomResourceDefinition {
-	return crd("MeshPolicy", "MeshPolicies", crdConfigs[Authentication], "", "", "", extensionsobj.ClusterScoped)
+	return nil
 }
 
-func crd(kind string, plural string, config crdConfig, appLabel string, pckLabel string, istioLabel string, scope extensionsobj.ResourceScope) *extensionsobj.CustomResourceDefinition {
-	return crdL(kind, plural, config, appLabel, pckLabel, istioLabel, scope, false)
-}
+func (r *CRDReconciler) load(f io.Reader) error {
+	var b bytes.Buffer
 
-func crdL(kind string, plural string, config crdConfig, appLabel string, pckLabel string, istioLabel string, scope extensionsobj.ResourceScope, list bool) *extensionsobj.CustomResourceDefinition {
-	singularName := strings.ToLower(kind)
-	pluralName := strings.ToLower(plural)
-	labels := make(map[string]string)
-	if len(appLabel) > 0 {
-		labels["app"] = appLabel
-	}
-	if len(pckLabel) > 0 {
-		labels["package"] = pckLabel
+	var yamls []string
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if line == "---" {
+			yamls = append(yamls, b.String())
+			b.Reset()
+		} else {
+			if _, err := b.WriteString(line); err != nil {
+				return err
+			}
+			if _, err := b.WriteString("\n"); err != nil {
+				return err
+			}
+		}
 	}
-	if len(istioLabel) > 0 {
-		labels["istio"] = istioLabel
+	if s := strings.TrimSpace(b.String()); s != "" {
+		yamls = append(yamls, s)
 	}
-	crd := &extensionsobj.CustomResourceDefinition{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:   fmt.Sprintf("%s.%s.istio.io", pluralName, config.group),
-			Labels: labels,
-		},
-		Spec: extensionsobj.CustomResourceDefinitionSpec{
-			Group:   fmt.Sprintf("%s.istio.io", config.group),
-			Version: config.version,
-			Versions: []extensionsobj.CustomResourceDefinitionVersion{
-				{
-					Name:    config.version,
-					Served:  true,
-					Storage: true,
-				},
-			},
-			Scope: scope,
-			Names: extensionsobj.CustomResourceDefinitionNames{
-				Plural:     pluralName,
-				Kind:       kind,
-				Singular:   singularName,
-				Categories: config.categories,
-			},
-		},
+
+	for _, yaml := range yamls {
+		s := json.NewYAMLSerializer(json.DefaultMetaFactory, scheme.Scheme,
+			scheme.Scheme)
+
+		obj, _, err := s.Decode([]byte(yaml), nil, nil)
+		if err != nil {
+			continue
+		}
+
+		var crd *extensionsobj.CustomResourceDefinition
+		var ok bool
+		if crd, ok = obj.(*extensionsobj.CustomResourceDefinition); !ok {
+			continue
+		}
+
+		crd.Status = extensionsobj.CustomResourceDefinitionStatus{}
+		crd.SetGroupVersionKind(schema.GroupVersionKind{})
+		labels := crd.GetLabels()
+		labels[createdByLabel] = "istio-operator"
```

Comment:
Could we move "istio-operator" to a const var?

Reply:
👍